### ISSUE TITLE
pbench-uperf does not collects UDP data on rx side

### ISSUE THREAD

**Issue by vikaschoudhary16** [distributed-system-analysis/pbench#497]
pbench-uperf collects results from tx side, which are not helpful and gives wrong perception. It should be collected on rx side.

**Comment by portante** [comment]
@atheurer, this seems like it would be a real problem, if it is true.  But I am not sure about that given we been running it so long this way.

**Comment by vikaschoudhary16** [comment]
These are the [results](http://pbench.perf.lab.eng.bos.redhat.com/results/ose-master/uperf_pod-to-pod-OCP-SDN_PODS_1_UDP_2017-02-28_07:15:42/result.html) from the container running inside a VM, where physical link is 10 GIG. For 16 KB msg size more than 28 Gb/sec.

Another such [results](http://pbench.perf.lab.eng.bos.redhat.com/results/ose-master/uperf_OCP-SDN_1450_pod-to-pod-MN_PODS_1_UDP_2017-02-17_13:32:19/result.html). Here also same setup as above(10 GIG link speed).

This problem seems to be there from a very long time. Please take a look at the Note about UDP results in this [report](https://mojo.redhat.com/docs/DOC-1044937#jive_content_id_Results) which is prepared more than 6 months back.

**Comment by vikaschoudhary16** [comment]
Ok, i understood the root cause and the fix as well. Issue is in parsing and post-processing uperf-client-output.txt (below is a sample). [uperf-postprocess](https://github.com/distributed-system-analysis/pbench/blob/master/agent/bench-scripts/postprocess/uperf-postprocess) generates a csv file from all the (time, value) tuples which are for the sender side and thus gives incorrect result in UDP case.

At the end of output.txt (in the bold below), there are throughput numbers from server side also and in UDP case that should be used to generate csv file.

> timestamp_ms:1488862304305.5793 name:Txn2 nr_bytes:323748364288 nr_ops:19760032
> timestamp_ms:1488862305306.2229 name:Txn2 nr_bytes:335893495808 nr_ops:20501312
> timestamp_ms:1488862306306.5054 name:Txn2 nr_bytes:348396716032 nr_ops:21264448
> timestamp_ms:1488862307307.5769 name:Txn2 nr_bytes:360801894400 nr_ops:22021600
> timestamp_ms:1488862308308.5132 name:Txn2 nr_bytes:373143371776 nr_ops:22774864
> timestamp_ms:1488862309309.5930 name:Txn2 nr_bytes:385629552640 nr_ops:23536960
> timestamp_ms:1488862310310.5552 name:Txn2 nr_bytes:397988069376 nr_ops:24291264
> timestamp_ms:1488862311311.6357 name:Txn2 nr_bytes:409155403776 nr_ops:24972864
> timestamp_ms:1488862312512.8682 name:Txn2 nr_bytes:420249862144 nr_ops:25650016
> 
> timestamp_ms:1488862312512.9817 name:Txn3 nr_bytes:0 nr_ops:0
> timestamp_ms:1488862312512.9958 name:Txn3 nr_bytes:0 nr_ops:0
> 
> -------------------------------------------------------------------------------
> timestamp_ms:1488862312613.2476 name:Total nr_bytes:420249862144 nr_ops:25650020
> Group Details
> -------------------------------------------------------------------------------
> timestamp_ms:1488862312513.0471 name:Group0 nr_bytes:0 nr_ops:1
> 
> 
> Strand Details
> -------------------------------------------------------------------------------
> timestamp_ms:1488862312513.0479 name:Thr0 nr_bytes:101359026176 nr_ops:6186467
> timestamp_ms:1488862312513.0479 name:Thr0 nr_bytes:101359026176 nr_ops:6186467
> timestamp_ms:1488862312513.0479 name:Thr0 nr_bytes:101359026176 nr_ops:6186467
> timestamp_ms:1488862312513.0479 name:Thr0 nr_bytes:101359026176 nr_ops:6186467
> 
> 
> Txn                Count         avg         cpu         max         min 
> -------------------------------------------------------------------------------
> Txn0                   4     48.20us      0.00ns     63.42us     41.11us 
> Txn1             1603128     89.81us      0.00ns     12.37ms     64.78us 
> Txn2                   4      2.12us      0.00ns      2.41us      1.75us 
> 
> 
> Flowop             Count         avg         cpu         max         min 
> -------------------------------------------------------------------------------
> connect                4     47.82us      0.00ns     62.95us     40.72us 
> write           25650016      5.60us      0.00ns     12.37ms     64.65us 
> disconnect             4      1.39us      0.00ns      1.52us      1.18us 
> 
> 
> Netstat statistics for this run
> -------------------------------------------------------------------------------
> Nic       opkts/s     ipkts/s      obits/s      ibits/s
> eth0       668982           1    87.91Gb/s     1.19Kb/s 
> -------------------------------------------------------------------------------
> 
> Run Statistics
> Hostname            Time       Data   Throughput   Operations      Errors
> -------------------------------------------------------------------------------
> **10.128.0.5        38.34s     7.01GB    _`1.57Gb/s`_       459586        0.00**
> **master            38.34s   391.39GB    _`87.68Gb/s`_     25650028        0.00**

cc @portante  @atheurer @jeremyeder

**Comment by portante** [comment]
@vikaschoudhary16, @atheurer, we have to verify what uperf means by those numbers.  In an IRC conversation, @vikaschoudhary16 said `iperf` verifies the 1.57Gb/s number.  And since he only has 40Gb/s links used for this result, 87.68 Gb/s does not seem possible.

**Comment by atheurer** [comment]
I agree with the above statements regarding the 1.57Gb/s number.  However, I generally don't like using UDP for stream type tests because there is no flow control.  Is there a particular need for UDP over TCP?

I suppose we could add another metric to our existing results for the server-side stats.  We don't get the stats over time as we do with the client, but it's better than nothing.  Perhaps if someone really wanted the server stats over time, uperf itself could be modified to output those from the server stdout (may require a bit of c coding).

**Comment by vikaschoudhary16** [comment]
@atheurer Agree that default tuning in linux is not for higher throughput rather it is for low latency.

Just to summarize may be we can break this into one bug fix and two enhancement PRs(one in pbench-agent and another in uperf):
1. Minimum first step should be to change description of udp stream results to indicate that these are client side stats.
2. Enhancement in pbench-agent to show additional server side stats as well can be 2nd step.
3. And then final and third could be to enhance uperf code as you mentioned to show server stats over time.

**Comment by atheurer** [comment]
Agreed, although for step1, all current stats should be changed to client stats
